add real world phoenix kubernetes post (#110)
kabisa/theguild.nl: draft/real-world-phoenix-lets-deploy-using-kubernetes.md

# Real World Phoenix |> Let's D4Y |> __using__ K8S

> As promised in my [last post](https://www.theguild.nl/real-world-phoenix-lets-d4y/) here is step 3 in my deployment adventure. The one

Reviewer: ..., here is step 3... (comma)

where I use Kubernetes!

In the last post we explored a very straightforward and simple
way to deploy a Phoenix application using the built-in Elixir releases on render.com. 

Today I want to discuss a much more complicated setup, using Kubernetes. I know for
most applications you will surely not need a setup using Kubernetes, but I've
been having an urge to try it out for a while now and wanted to see if it is
really that complicated as I seem to be hearing through the grapevine. Who
knows, maybe it turns out it's actually not that hard at all. 
Also I really like the fact that I can have a cluster to spin up nodes (ie.
applications) of all the experiments / ideas / apps that I'm currently playing
with, so I can actually expose those into the Real World with a reproducable and
easy deployment workflow. Let's see if we can make that dream come true! 

## The Kubernetes Cluster

> If you are not aware of what Kubernetes is, head over to [their website](https://kubernetes.io/docs/tutorials/kubernetes-basics/), as they have great documentation to get started on the basics.

Reviewer: 👍

To use Kubernetes for our deployment the first thing we'll need is a Kubernetes
cluster. This is the part where you see most tutorials and guides grab for a
local setup with minikube. While I like that it's possible to do this, it still
seems a bit too far from real life for me, so I wanted to have a bit more of a
production-like environment, like... for instance... a production environment :) 

The beauty is that these days there are more and more managed Kubernetes
services that do all the hard work like setting up the cluster, managing

> upgrades etc etc. I'll be using the service DigitalOcean is offering, because I

Reviewer: You could opt to split the sentence here at `because`. (lange zin)

# Real World Phoenix |> Let's D4Y |> __using__ K8S

As promised in my last post here is step 3 in my deployment adventure. The one
where I use Kubernetes!

In the last post we explored a very straightforward and simple
way to deploy a Phoenix application using the built-in Elixir releases on render.com. 

Today I want to discuss a much more complicated setup, using Kubernetes. I know for
most applications you will surely not need a setup using Kubernetes, but I've
been having an urge to try it out for a while now and wanted to see if it is
really that complicated as I seem to be hearing through the grapevine. Who
knows, maybe it turns out to not to be that hard in the end. 
Also I really like the fact that I can have a cluster to spin up nodes (ie.
applications) of all the experiments / ideas / apps that I'm currently playing
with, so I can actually expose those into the Real World with a reproducable and
easy deployment workflow. Let's see if we can make that dream come true! 

## The Kubernetes Cluster

> If you are not aware of what Kubernetes is (under what rock have you been

Reviewer 2: ```suggestion
If you are not aware of what Kubernetes is
```

Maybe let's not insult readers?

hiding...?), head over to [their
website](https://kubernetes.io/docs/tutorials/kubernetes-basics/), 
as they have great documentation to get started on the basics.

To use Kubernetes for our deployment the first thing we'll need is a Kubernetes
cluster. Now this is the part where you see most tutorials and guides grab for a
local setup with minikube. While I like that it's possible to do this, it still
seems a bit too far from real life for me, so I wanted to have a bit more of a
production-like environment, like... for instance... a production environment :) 

> The beauty is that these days there are more and more managed kubernetes

Reviewer 2: ```suggestion
The beauty is that these days there are more and more managed Kubernetes
```

services that do all the hard work like setting up the cluster, managing

> upgrades etc etc. I'll be using the service digitalocean is offering, because I

Reviewer 2: ```suggestion
upgrades etc etc. I'll be using the service DigitalOcean is offering, because I
```

> already have a lot of content running at digitalocean, so it is known territory

Reviewer 2: ```suggestion
already have a lot of content running at DigitalOcean, so it is known territory
```

for me and they make it really easy to get sarted with Kubernetes.

> If you haven't used digitalocean before, you can use this [referral

Reviewer 2: ```suggestion
If you haven't used DigitalOcean before, you can use this [referral
```

link](https://m.do.co/c/78020d21b236) to create an account and get $100 credit
to spend in first 60 days. So more than enough to follow along with this tutorial.

## Terraform + Helm

Oh and we are going to need to use 2 more tools to get this setup up and

> running. Terraform and Helm. We'll use terraform to setup most of the infrastructure.

Reviewer 2: Maybe link to Terraform and Helm here?

few of the tools needed, so reverted to Helm (the kubernetes package manager) to
install Traefik and our Gitlab Runner. 

Our Terraform setup will include:
1. Setting up a DigitalOcean Kubernetes Cluster with 2 worker nodes
2. Connecting a gitlab project to the kubernetes cluster
3. Creating a kubernetes_secret to be able to pull images from our private gitlab
   registry.
4. Setup a managed database cluster service @ digitalocean

And we'll use Helm to:
1. Install Traefik as an ingress controller
2. Create a Gitlab Runner in our cluster 

So let's get crackin'!

### install doctl to talk to your digitalocean account

If we want to automate any of this stuff we'll have to be able to talk to
digitalocean programmatically, so installing doctl is step one. Please refer to
this guide to get that setup: [doctl up and running
guide](https://blog.digitalocean.com/introducing-doctl/)

After you have `doctl` installed, make sure you create a personal access token
in digitalocean and add that in an environment variable and also initialize your
account.
```bash
export DIGITALOCEAN_TOKEN=[token]
doctl auth init -t $DIGITALOCEAN_TOKEN
doctl get account
```

### Install Terraform

> Ofcourse there is a handy install guide for terraform also. Actually just an

Reviewer 3: ```suggestion
Of course there is a handy install guide for terraform also. Actually just an
```

executable, as it is written in go, so they have a handy packaged binary you can use.
See: [terraform install](https://www.terraform.io/downloads.html) 

### Setup our project

If you want to follow along with this guide, create a folder that will hold your
terraform configs and create a `main.tf` file that will hold our terraform declarations.

```bash
mkdir kube-terra
cd kube-terra
touch main.tf
```

### Setup the digitalocean Kubernetes cluster

Terraform works in a declaritive manner, which means that we state what we want
the world to be like and terraform figures out how to get to that state. In
terraform you describe an item you want to exist as a resource. A Kubernetes
cluster is a form of a resource that is provided by one of the many providers
that exist in terraform. So let's create a digitalocean kubernetes cluster:

Creating a resource is always in the format: `resource [kind] [reference] {}`
We can use the reference we provide later on to refer back to this resource.

> ```json

Reviewer 3: ```suggestion
```hcl
```

Author: ahh yes, that was it. Thanks!

resource "digitalocean_tag" "kubernetes-cl01" {
  name = "kubernetes-cl01"
}

resource "digitalocean_kubernetes_cluster" "cl01" {
  name    = "cl01"
  region  = "ams3"

>   version = "1.16.2-do.1"

Reviewer 3: This version was no longer available for me. I did have success using the `digitalocean_kubernetes_versions` resource to get the latest 1.16 as described here: https://www.terraform.io/docs/providers/do/d/kubernetes_versions.html

  node_pool {
    name       = "default"
    size       = "s-1vcpu-2gb"
    node_count = 2
    tags       = ["${digitalocean_tag.kubernetes-cl01.id}"]
  }
}
```

Now you can get your cluster setup by issuing the following commands in the folder you
created:

```bash
tarraform init
terraform apply
```

This will take about 5 minutes and then you'll have a fresh kubernetes cluster
up and running. Now that was easy!

### Install Traefik

> Kubernetes needs an ingress controller to route traffik from the outside world

Reviewer 4: traffic

to the services that are running inside the cluster as these services are not

> exposed to the outside world, which is a good thing! Nginx is often used, but I went with Traefik, a very nice alternative that has a lot of

Reviewer 2: Isn't it easier to ease DO load balancers?

Author: Traeffik actually uses a DigitalOcean LoadBalancer

Author: I'll add that information in

Reviewer 2: Are you sure?

nice functinality out of the box, like automatic ssl certificates using Let's Encrypt
and auto discovery of services running in the cluster.

Create a file called `traefik-values.yml` in the root of your kubernetes config

> directory and add the following content. If you manage dns settings with

Reviewer 4: FYI: using DNS as your acme provider also allows you to get a valid certificate when your production is behind a firewall.

```yaml

image: traefik
dashboard:
  enabled: true
  domain: domain.example.com                     # put a (sub)domain here where you want to access the traefik dashboard
serviceType: LoadBalancer
rbac:
  enabled: true
# ssl:
#   enabled: true                                # Enables SSL
#   enforced: true                               # Redirects HTTP to HTTPS
# acme:
#   enabled: true                                # Enables Let's Encrypt certificates
#   staging: false                               # Use Lets Encrypt staging area for this example. For production purposes set this to false
#   email: info@drumusician.com                  # Email address that Let's Encrypt uses to notify about certificate expiry etc.
#   challengeType: "dns-01"
#   dnsProvider:
#     name:  digitalocean                        # This is why you need your domain to be under Digital Ocean control
#     digitalocean:
#       DO_AUTH_TOKEN: $DIGITALOCEAN_TOKEN
#   domains:
#     enabled: true
#     domainsList:
#       - main: domain.example.com               # domain that belongs to this certificate
```

Now we can go ahead and install traefik in our cluster:

Install helm:
```bash

> brew install kubernetes-helm

Reviewer 4: TIL: 
> The kubernetes-helm homebrew formula is a community-owned project

```
Warning: helm 3.0.2 is already installed and up-to-date
To reinstall 3.0.2, run `brew reinstall helm`
➜  ~ brew install helm
Warning: helm 3.0.2 is already installed and up-to-date
To reinstall 3.0.2, run `brew reinstall helm`
```

```

Pull in our fresh cluster configuration locally (otherwise helm will
fail):
```bash
doctl kubernetes cluster kubeconfig save cl01
```

And install traefik:

```bash

> helm install traefik --values traefik-values.yml stable/traefik

Reviewer 3: I had to add the Helm stable charts repo first (see https://helm.sh/docs/intro/quickstart/):

```bash
helm repo add stable https://kubernetes-charts.storage.googleapis.com/
```

```

After Traefik is installed we have enough running to be able to deploy our stuff
to the cluster. We are going to use gitlab-ci to deploy so we'll need to
install a gitlab-runner in our cluster first so we can use that in our project.

### Creating a kubernetes secret to access our registry

> If we install the gitlqb runner in the next step, we'll need a kubernetes secret

Reviewer 3: ```suggestion
If we install the gitlab runner in the next step, we'll need a kubernetes secret
```

that will be able to pull from our private registry in gitlab.
Go to https://gitlab.com/profile/personal_access_tokens and create an access

> token to access the gitlab registry. Scopes: appi, read_registry.

Reviewer 3: ```suggestion
token to access the gitlab registry. Scopes: api, read_registry.
```

I suppose.

Then create a file at `~/.docker/docker-registry.json` with the credentials for
the access token:
```json
{
  "registry.gitlab.com": {
    "username": "",            # access-token-name
    "password": "",            # token
    "email": ""                # email of the gitlab account
  }
}
```

And then in our main.tf we can create the kubernetes secret.

> ```json

Reviewer 3: ```suggestion
```hcl
```

add the secret:
```bash
terraform init
terraform apply
```

### Install Gitlab Runner

For this setup to work we will need our runner to run inside of our cluster
and have enough access rights to actually spin-up pods and services in our
cluster. So we'll install that using helm as well. Gitlab has an install script
from their interface, but that doesn't provide you with customisation options
that we'll need for our use case.

Create the following file in your kubernetes dir: `gitlab-runner-values.yml` and
add this content:

```yaml
gitlabUrl: https://gitlab.com/
runnerRegistrationToken: "" # copy this from your gitlab project settings: 
rbac:
  create: true
  clusterWideAccess: true
runners:
  privileged: true
imagePullSecrets: ["gitlab.com"]
```

Now let's install the runner:

```bash
helm install --namespace default gitlab-runner  -f gitlab-runner-values.yml gitlab/gitlab-runner
```

### Connect a gitlab project to the cluster

This is something you can do through the gitlab interface, but I like automating
it here as well. We need to provide the gitlab_project in terraform like this.
If you don't have a project yet, you should create it on gitlab.com and add the
ID here.

> ```

Reviewer 3: ```suggestion
```hcl
```

data "gitlab_project" "project-x" {
  id = [your-project-id]
}
```

Then we can use this reference to create the settings in our project:

> ```

Reviewer 3: ```suggestion
```hcl
```

  project                       = data.gitlab_project.project-xid
  name                          = "my-awesome-cluster"
  domain                        = "[mydomain.com]"
  enabled                       = true
  kubernetes_api_url            = digitalocean_kubernetes_cluster.cl01.endpoint
  kubernetes_token              = digitalocean_kubernetes_cluster.cl01.kube_config[0].token
  kubernetes_ca_cert            = base64decode(digitalocean_kubernetes_cluster.cl01.kube_config[0].cluster_ca_certificate)
  kubernetes_namespace          = ""
  kubernetes_authorization_type = "rbac"
  environment_scope             = "*"
}
```

And lastly we are not going the skip the database, as that is too easy to do.
You could potentially setup a database persistent volume in kubernetes and have
db_pods spinup, but I think that it is much easier to have the
database as a separate service outside of kubernetes. The digitalocean managed
database service is a great option. You pay a little extra, but it takes care of
backups etc. It is just one of those things that you don't want to worry about,
right?

Ofcourse, we can easily add the creation of a database cluster to our terraform
setup:

> ```json

Reviewer 3: ```suggestion
```hcl
```

resource "digitalocean_database_cluster" "postgres-db" {
  name       = "postgres-db-cluster"
  engine     = "pg"
  version    = "11"
  size       = "db-s-1vcpu-1gb"
  region     = "ams3"
  node_count = 1
}
```

And while we are at it, let's create the database and firewall settings as well.

> ```json

Reviewer 3: ```suggestion
```hcl
```

resource "digitalocean_database_db" "test-prod" {
  cluster_id = digitalocean_database_cluster.postgres-db.id
  name       = "test_prod"
}

resource "digitalocean_database_firewall" "db-fw" {
  cluster_id = digitalocean_database_cluster.postgres-db.id

  rule {
    type  = "k8s"
    value = digitalocean_kubernetes_cluster.cl01.id
  }
}
```

This concludes our cluster setup. To get our app deployed we'll have to add some

> setup to our project. We'll need to dockerize our project, add kubctl deploy

Reviewer 3: ```suggestion
setup to our project. We'll need to dockerize our project, add kubectl deploy
```

config files and add a gitlab-ci.yml that will trigger the gitlab-ci pipeline.

## A project to deploy

Now we have our kubernetes cluster up-and-running it is time to see how we would
deploy our app to the cluster.

I have a git tag prepared for you to use from my real_world_phoenix project. You
can clone that checkout to use to deploy to the cluster. It has the necessary
files to trigger deployment which I'll explain next.

Go ahead and clone the project from the tag I created:
```

> git clone git@gitlab.com:drumusician/real_world_phoenix.git --branch kubernetes-deploy

Reviewer 3: Should this still be pushed?

```zsh
$ git clone git@gitlab.com:drumusician/real_world_phoenix.git --branch kubernetes-deploy
Cloning into 'real_world_phoenix'...
remote:
remote: ========================================================================
remote:
remote: You are not allowed to download code from this project.
remote:
remote: ========================================================================
remote:
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Author: ah maybe, I'll check

Author: @rhynix does this url work: https://gitlab.com/drumusician/real_world_phoenix.git

  CONTAINER_RELEASE_IMAGE: $REGISTRY/$CI_PROJECT_NAMESPACE/$CI_PROJECT_NAME:$CI_COMMIT_SHORT_SHA

stages:
  - init
  - build
  - deploy

# the dot syntax makes this a hidden step that we can include in other places further down.
.elixir_default: &elixir_default
  image: elixir:1.9
  before_script:
    - mix local.hex --force
    - mix local.rebar --force

.javascript_default: &javascript_default
  image: node:alpine
  before_script:
    - cd assets

# Compile our elixir artifacts
elixir_compile:
  <<: *elixir_default
  stage: init
  script:
    - mix deps.get --only test
    - mix compile
    - mix compile --warnings-as-errors
  artifacts:
    paths:
      - mix.lock
      - _build
      - deps

# compile javascript artifacts
javascript_deps:
  <<: *javascript_default
  stage: init
  script:
    - npm install --progress=false
  artifacts:
    paths:
      - assets/node_modules

# build our container image
build:
  stage: build
  image:
    name: gcr.io/kaniko-project/executor:debug
    entrypoint: [""]
  script:
    - echo "{\"auths\":{\"$CI_REGISTRY\":{\"username\":\"$CI_REGISTRY_USER\",\"password\":\"$CI_REGISTRY_PASSWORD\"}}}" > /kaniko/.docker/config.json

>     - /kaniko/executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/Dockerfile --destination $CONTAINER_RELEASE_IMAGE

Reviewer 4: oooh Kaniko. Me like

# and deploy our kubernetes cluster
deploy:
  stage: deploy
  image:
    name: lwolf/kubectl_deployer:latest
  script:

>     - cat deploy.yml | envsubst | kubectl apply -f -

Reviewer 4: TIL: `envsubst`

kind: Service
metadata:
  name: real-world-phoenix-service
spec:
  ports:
  - name: web
    port: 8001
    protocol: TCP
  selector:
    app: real_world_phoenix
```

* ingress.yml
The mapping of our domain name to our service

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: real-world-phoenix-ingress
spec:
  rules:
  - host: $DOMAIN
    http:
      paths:
      - path: "/"
        backend:
          serviceName: real-world-phoenix-service
          servicePort: web
```

* deploy.yml
Our deployment configuration

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: real-world-phoenix
  labels:
    app: real_world_phoenix
spec:
  revisionHistoryLimit: 5
  replicas: 1
  selector:
    matchLabels:
      app: real_world_phoenix
  template:
    metadata:
      labels:
        app: real_world_phoenix
    spec:
      imagePullSecrets:
        - name: "gitlab.com"
      containers:
      - name: real-world-phoenix
        image: registry.gitlab.com/$CI_PROJECT_NAMESPACE/$CI_PROJECT_NAME:$CI_COMMIT_SHORT_SHA
        imagePullPolicy: Always
        name: real-world-phoenix-deployment
        resources: {}
        ports:
          - containerPort: 8001

>         env:

Reviewer 4: Nitpick: If it's this big, I like using env vars from configmap.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: k8s.gcr.io/busybox
      command: [ "/bin/sh", "-c", "env" ]
      envFrom:
      - configMapRef:
          name: special-config
  restartPolicy: Never
```

Author: ah yes, that is one part I haven't explored yet.

          - name: SECRET_KEY_BASE
            value: $SECRET_KEY_BASE
          - name: DB_USER
            value: $DB_USER

>           - name: DB_PASSWORD

Reviewer 4: Passwords should come from secrets even though secrets do not provide security.

Author: ? Is there something I should do differently here? Gitlab does mask these variables as well in case something gets logged.

Reviewer 4: If you do `kubectl get pod <your-pod>` it would show your password in plain text. I would create a secret and ad a secretref with which you pull in all variables in the secret as env vars.

Author: ah ok, I'll look into that some more, thanks!
